documentation/API-GUIDELINES.md

```diff
@@ -15,16 +15,27 @@ The following paragraphs contain additional recommendations.
 
 ## Construction and Destruction of Drivers
 
-- Drivers take peripherals and pins via the `PeripheralRef` pattern - they don't consume peripherals/pins.
+- Drivers should take peripherals via the `PeripheralRef` pattern - they don't consume peripherals directly.
+- If a driver requires pins, those pins should be configured using `fn with_pin_name(self, pin: impl Peripheral<P = InputConnection> + 'd) -> Self)` or `fn with_pin_name(self, pin: impl Peripheral<P = OutputConnection> + 'd) -> Self)`
+- If a driver supports multiple peripheral instances:
+  - The peripheral instance type should be positioned as the last type parameter of the driver type.
+  - The peripheral instance type should default to a type that supports any of the peripheral instances.
+  - The author is encouraged to use `crate::any_peripheral` to define the "any" peripheral instance type.
+  - The driver should implement a `new` constructor that automatically converts the peripheral instance into the any type, and a `new_typed` that preserves the peripheral type.
+- If a driver implements both blocking and async operations, or only implements blocking operations, but may support asynchronous ones in the future, the driver's type signature should include a `crate::Mode` type parameter.
+- By default, constructors should configure the driver for blocking mode. The driver should implement `into_async` (and a matching `into_blocking`) function that reconfigures the driver.
+  - `into_async` should configure the driver and/or the associated DMA channels. This most often means enabling an interrupt handler.
+  - `into_blocking` should undo the configuration done by `into_async`.
+- The asynchronous driver implemntation should also expose the blocking methods (except for interrupt related functions).
 - Consider adding a `Drop` implementation resetting the peripheral to idle state.
-- Consider using a builder-like pattern for configuration which must be done during initialization.
```

Comment:
I noticed this was removed, is this in favour of the `If a driver requires pins` entry? If so I don't feel like they're fully equivalent, look at SPI for example we use a builder like pattern to set more than just pins on the peripheral.

Reply:
Oh I thought this is some `Config::with_xy` thing.

Reply:
Oh I see, reading that again it could read like that - shall we clarify it to "Consider using a builder-like pattern for driver construction."?

Reply:
That's the thing, I'm not sure we should force this. We can have the rule because of the status quo, but:

- We might want to prefer a separate `SpiDma::new` constructor instead of `Spi::with_dma`, so we can ensure that the DMA channel really is compatible with the SPI peripheral. I'm not sure we'll do this, because only PDMA devices have this issue, but it would actually be more consistent with I2S, for example.
- There are arguments in favour of `set_pin` instead of `with_pin` (reconfigurability).
- What else is there that we might want to write as a builder method?

Reply:
> * because only PDMA devices have this issue

And the P4 😛

Reply:
> This is still a builder-like pattern but with &mut self instead of self, right?

I wouldn't consider it as one. While it's possible to chain `fn x(&mut self) -> &mut Self` functions, you no longer have an owned driver at the end.

Reply:
> > * because only PDMA devices have this issue
> 
> And the P4 😛

No, the P4 behaves well enough with the current scheme - a peripheral either takes an AHB GDMA channel, or an AXI one, but there is no crosstalk between them. The problem with PDMA chips is that I2s0 and I2s1 DMA channels erase into the same type but aren't cross-compatible between peripherals.

Reply:
> I wouldn't consider it as one. While it's possible to chain fn x(&mut self) -> &mut Self functions, you no longer have an owned driver at the end.

In which case, if we want the ergonomics we have now with the `with_x` API (IMO we do, it's nice and avoid huge constructors), they _need_ to be self. We can always add `set_` variants of the same API which take `&mut self` instead to address runtime reconfig?

Reply:
I think, as we don't explicitly support it in esp-hal (because it kind of needs unsafely cloning pins), I'd direct users who need the functionality towards `replace_with`. They can then implement a SPI wrapper that has a Config struct that contains pins, and they can implement SetConfig with replace_with to apply the pins. `set_` is nice but not required.

Reply:
Okay I guess discounting `with_dma` we're good here?

Reply:
We can't return the pins as we forget them, and in a SetConfig API it would also be funny that the pins would rotate between devices. I'm saying `[maybe?]` because I'm unsure what happens if the configs contain static references to pins, it might end up working without unsafe cloning.

Reply:
> We can't return the pins as we forget them

Not completely, the pin would be recorded in the signal register, which we can read later.

Reply:
The hardware allows routing a peripheral output signal to multiple pins, so it won't be unambiguous in general.

That being said, we have a bug where we don't clear output connections if you bind a new pin, we'll have to resolve that otherwise shared I2C like this won't be happy.

Reply:
I think the `set_pin` API or generally runtime pin replaces warrants it's own discussion (I encourage to start this with an RFC). For now I think `with_x` taking `self` is a nice pattern that we will continue to use regardless of the outcome of the runtime pin replacement RFC.